File: core-js/src/elements/RigidBody.ts
import { Matrix4x4 } from '../math/Matrix4x4';
import { Node } from './Node';

export class RigidBody {
  public id: string;
  /** The absolute transformation of this rigid body in world space */
  public transform: Matrix4x4;
  /** Nodes that are rigidly attached to this body */
  public nodes: Map<string, Node> = new Map();

  constructor(id: string) {
    this.id = id;
    this.transform = new Matrix4x4();
  }

  addNode(node: Node): void {
    this.nodes.set(node.id, node);
  }

  /**
   * Recalculates absoluteTransform for all attached nodes
   * using the formula: NodeAbs = BodyTransform * NodeLocal
   */
  updateNodes(): void {
    for (const node of this.nodes.values()) {
      node.absoluteTransform = this.transform.multiply(node.localTransform);
    }
  }
}

File: core-py/src/wigglesolve/elements/rigid_body.py
from __future__ import annotations
from wigglesolve.math.matrix4x4 import Matrix4x4
from wigglesolve.elements.node import Node


class RigidBody:
    def __init__(self, id: str) -> None:
        self.id = id
        self.transform = Matrix4x4()
        self.nodes: dict[str, Node] = {}

    def add_node(self, node: Node) -> None:
        self.nodes[node.id] = node

    def update_nodes(self) -> None:
        for node in self.nodes.values():
            node.absolute_transform = self.transform.multiply(node.local_transform)
